### ISSUE TITLE
Alternative for asyncspinner in ROS2

### ISSUE THREAD

**Issue by pankhurivanjani** [TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani#6]
Referring to issue in [#Issue5](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/issues/5)

Without it, it won't be possible to use ROS2 with camViz.

Links to multithreaded executor in ROS2: 

* [rclcpp::executors::multi_threaded_executor::MultiThreadedExecutor Class Reference](http://docs.ros2.org/beta3/api/rclcpp/classrclcpp_1_1executors_1_1multi__threaded__executor_1_1MultiThreadedExecutor.html)

* [rclcpp::executor::Executor Class Reference](http://docs.ros2.org/beta1/api/rclcpp/classrclcpp_1_1executor_1_1_executor.html)
***********************************************************************************************************
From one of the github issues [support for AsyncSpinner](https://github.com/ros2/rclcpp/issues/335) I found this:
For ros2 dashing:

```
using rclcpp::executors::MultiThreadedExecutor;
MultiThreadedExecutor executor;
executor.add_node(xxx_node);
```
***********************************************************************************************************

I have posted this there too.

I am having issues in implementing it in ROS dashing.
With ROS1 I was using start and stop. What to do in this case?
Also if I have:

```
lc2->start();
std::cout << "hi";
 std::cout << "hey"; 
```

where start:

```
void start(){
 RCLCPP_INFO(this->get_logger(), "starting");
this->executor->spin();}
```

My code is not printing "hi" and "hey". It is not moving to further lines wheras with ROS since I used asyncspinner I had no problem.

***********************************************************************************************************

I studied more about single and multi-threaded nodes and use of various spins in ROS and in useful link [ros-spins](https://yuzhangbit.github.io/tools/several-ways-of-writing-a-ros-node/). I found this. 

```
ros::spinOnce() // ask ROS to handle callbacks  
doYourWork()    // this line will be executed
```

```
ros::spin()  // ask ROS to handle callbacks
youWantToDoYourWork()  // this line will never be executed.
```
 

Maybe this is what is happening with me. Print statements which I wrote after starting were not executed.

**AsyncSpinner** provides more control to users than MultiThreadedSpinner, which is **similar to spinOnce()**. If you want to update the data of the node, you may need to put a mutex in your callback function. Another thing you need to note is that you need ros::wait for shutdown() after the spinner.start() for AsyncSpinner, or the ros node will only spin once.

**Comment by vinay0410** [comment]
Hi @pankhurivanjani ,
First, why do you need to print something after starting the service ?
start should be the last call, since it is a blocking call, and you won't move forward until the end of the process (that too only if you have callbacks for SIGTERM)

It was working in AsyncSpinner, since the start process was started asynchronously and therefore wasn't a blocking call.

**Comment by vinay0410** [comment]
Also, is the following code not working ?
```
using rclcpp::executors::MultiThreadedExecutor;
MultiThreadedExecutor executor;
executor.add_node(xxx_node);
```

**Comment by vinay0410** [comment]
Also I don't think "spinOnce is not similar to asyncSpinner", it is very much different.
Asyncspinner starts the listener or publisher and then in a seperate simultaneously giving access to you current process thread, which enables execution of the line written below.

**Comment by pankhurivanjani** [comment]
Basically, my image subscriber is working fine ( I tested with cv::imshow) but I am getting no images in GUI. 

In this [camviz.cpp ](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camViz.cpp) I have this loop for displaying 

	while(viewer.isVisible()){
		//jderobot::ImageDataPtr data = camRGB->getImageData(format);
		rgb = camRGB->getImage();
		std::cout << "imageayo\n";

		viewer.display(rgb.data);
		viewer.displayFrameRate(0);
	
but after checking with gdb bugger and print statements I found my program was not going ahead after

camRGB = camViz::getCameraClient(argc,argv,server,topic,nodeName);

So I moved ahead in the file [ros1-ros2-interface ](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/mainsubcpp.hpp) I was calling start. After start we have :

client = (camViz::CameraClient*) lc2;

and getcameraclient would also be returning client for input to our GUI.

but due to blocking calls, it is not moving forwards to ```client = (camViz::CameraClient*) lc2;```

Printing statements were for checking this only since gdb bugger wasn't directing to issue.

**Comment by pankhurivanjani** [comment]
> Also I don't think "spinOnce is not similar to asyncSpinner", it is very much different.
> Asyncspinner starts the listener or publisher and then in a seperate simultaneously giving access to you current process thread, which enables execution of the line written below.

Okay!

In that blog I think it might be referring to thing that both give control to the user and not ros during spinning.

**Comment by pankhurivanjani** [comment]
> Also, is the following code not working ?
> 
> ```
> using rclcpp::executors::MultiThreadedExecutor;
> MultiThreadedExecutor executor;
> executor.add_node(xxx_node);
> ```

I have tried this code in my implementations:

[camsubros2exp.hpp](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/blob/master/camviz-updated/camsubros2exp.hpp)

It is giving no runtime errors but it was blocking a call I guess.

**Comment by vinay0410** [comment]
@pankhurivanjani , Are you sure it is a blocking call ?

**Comment by pankhurivanjani** [comment]
> @pankhurivanjani , Are you sure it is a blocking call ?

I am not really sure about it. Since I have kept the architecture of both ROS1 and ROS2 same, I compared the changes debugged the code and guessed it. 

I am not familiar much with multithreading so I read about these things recently.

**Comment by pankhurivanjani** [comment]
> @pankhurivanjani , Are you sure it is a blocking call ?

Is there any other way to know about the actual problem or check the blocking call thing?

**Comment by pankhurivanjani** [comment]
One more thing, I forgot to mention. 

In asynspinner we has start and stop.

But with this, I went through the documentation here [docs](http://docs.ros2.org/beta3/api/rclcpp/classrclcpp_1_1executors_1_1multi__threaded__executor_1_1MultiThreadedExecutor.html#ac82b0d01f97121ed59fd6cea15c534d8) and used spin.

It has this:

""""spin()
Do work periodically as it becomes available to us. Blocking call, may block indefinitely.""""

In my start too

**Comment by pankhurivanjani** [comment]
I was thinking that we may have to get new member functions like start and stop for this new 'executor' (namely spinner)

using rclcpp::executors::MultiThreadedExecutor;
MultiThreadedExecutor executor;
executor.add_node(xxx_node);

**Comment by pankhurivanjani** [comment]
Solved wrt [#8 ](https://github.com/TheRoboticsClub/colab-gsoc2019-Pankhuri_Vanjani/issues/8)
